tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.tags;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.immutables.value.Value;
+
+@Value.Immutable
+@Value.Style(
+        // Don't require GuavaModule
+        jdkOnly = true,
+        // detect 'get' and 'is' prefixes in accessor methods
+        get = {"get*", "is*"},
+        // Try to avoid leaking Immutable prefixed objects to avoid
+        // api fallout when we switch to another processor.
+        // Side effect: Return the base type from builders.
+        // Side effect: Nest implementation class inside of builder class.
+        visibility = Value.Style.ImplementationVisibility.PRIVATE,
+        builderVisibility = Value.Style.BuilderVisibility.PACKAGE,
+        // Default interface methods don't need to be annotated @Value.Default
+        defaultAsDefault = true)
+public abstract class TaggedMetric {
+
+    private static final List<String> DELIMITERS = Collections.unmodifiableList(Arrays.asList(":", ",", "[", "]"));
+
+    @Value.Parameter
+    abstract String name();
+
+    @Value.Parameter
+    abstract Map<String, String> tags();
+
+    /**
+     * Returns the canonical metric name including any tags.
+     *
+     * @return the canonical metric name
+     */
+    @Value.Lazy
+    public String canonicalName() {
+        if (nonEmptyTags().isEmpty()) {
+            return name();
+        } else {
+            StringBuilder builder = new StringBuilder();
+            builder.append(name()).append('[');
+
+            for (Iterator<Map.Entry<String, String>> it = nonEmptyTags().entrySet().iterator(); it.hasNext(); ) {
```

Comment:
I think you'll benefit at some point from sorting the keys (e.g. if you make a map of these things, you'll be able to lookup).

Reply:
like presumably the user is going to be like metricRegistry.meter(TaggedMetric.of("name", singletonMap("tag", "value"))

Reply:
applied sorting and some normalization of keys (lower casing, require <= 20 alpha numeric)